---
title: "Assessment 4"
author: "Julia Lay"
date: "2025-09-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##### First we load all packages needed mainly for Part 2 of this project. 
Note that the `count()` function must be distinguished within code chunks as it exists in both `seqinr` and `dplyr`.

```{r loading packages}
suppressPackageStartupMessages({
  library("seqinr") # is a package designed to process and analyse sequence data.
  library("R.utils") # general utilities like zip and unzip
  library("ggplot2") # for plotting graphs
  library("dplyr") # For group_by() and mutate() 
  library("patchwork") # to arrange plots
  library("forcats") # for fct_rev()
})
```


## Part 1: Importing files, data wrangling, mathematical operations, plots and saving code on GitHub

### Gene Expression Analysis

#### 1. Read in the file and show table with first six genes
The below reads in the file “gene_expression.tsv” making the gene identifiers the row names. The `head()` function in R then displays a table of values for the first six genes. By default, `head()` shows 6 rows but we could also specify the rows through specifying the number of rows i.e. `head(gene_data, n=6)`.

```{r}
# Read in data (tab separated)
gene_data <- read.table("gene_expression.tsv", header = TRUE, row.names = 1, sep = "\t")

# Show first six rows
head(gene_data)
```

#### 2. Add mean column
We make a new column which is the mean of the other columns. The first six genes is displayed again through the `head()` function.

```{r}
# Create new column with mean of other columns
gene_data$mean <- rowMeans(gene_data)

# Display first 6
head(gene_data)
```

#### 3. List top 10 genes with highest mean expression
We want to identify which genes have the highest average expression across all samples. The top 10 genes with the highest mean expression is shown in `top_genes`. The '-' sign before `gene_data$mean` tells it to sort in decreasing order.

```{r}
# Create variable to store top 10 genes with highest mean expression and sort in descending order
top_genes <- head(gene_data[order(-gene_data$mean), ], 10)

# Print out top genes
top_genes
```


#### 4. Determine the number of genes with a mean <10
Create variable `low_mean_count` and store the total number of mean values <10. When returning the `low_mean_count` value, we receive *35,988 genes* with a mean <10.

```{r}
#create variable for data with mean values less than 10
low_mean_count <- sum(gene_data$mean < 10)

#print out low mean count
low_mean_count
```

#### 5. Make a histogram plot of the mean values and include it into your report.
The below creates a histogram to visualise distribution of gene expression levels across all genes. A peak near lower values indicates many genes have low expression.

```{r}
# Create histogram using the mean column
hist(gene_data$mean, 
     breaks = 100, # Divide the x-axis into 100 bins
     xlim = c(0, 550000), # Set the range of the x-axis (mean expression values)
     ylim = c(0, 60000), # Set the range of the y-axis (frequency of genes)
     main = "Histogram of Mean Gene Expression", # Title of the histogram
     xlab = "Mean Expression") # Label for the x-axis
```


### Growth Data Analysis

#### 6. Import this csv file into an R object. What are the column names?
The column names are printed using `colnames()` as follows. They are: "Site", "TreeID", "Circumf_2005_cm", "Circumf_2010_cm", "Circumf_2015_cm", and "Circumf_2020_cm".

```{r}
#store csv into a variable
growth_data <- read.csv("growth_data.csv", header = TRUE)

#print out column names
colnames(growth_data)
```

#### 7. Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.
The sites available are either "northeast" or "southwest". Our columns suggest the start of the study period is 2005 and the end of the study is 2020. The below data frame contains a calculation of the mean and standard deviation for both sites at 2005 and 2020.

```{r}
#create a data frame for the statistics of mean and sd
stats <- data.frame(
  Site = unique(growth_data$Site),
  mean_2005 = tapply(growth_data$Circumf_2005_cm, growth_data$Site, mean),
  sd_2005   = tapply(growth_data$Circumf_2005_cm, growth_data$Site, sd),
  mean_2020 = tapply(growth_data$Circumf_2020_cm, growth_data$Site, mean),
  sd_2020   = tapply(growth_data$Circumf_2020_cm, growth_data$Site, sd)
)

#print out data frame for statistics containing the mean and sd
stats
```

#### 8. Make a box plot of tree circumference at the start and end of the study at both sites.
The box plot clearly shows that the tree circumference at the end for both sites grew, with the northeast site showing more growth than the southwest site at the end of the study in 2020.
```{r}
# Create combined vector of circumference values
circumference <- c(growth_data$Circumf_2005_cm, growth_data$Circumf_2020_cm)

# Create corresponding Site labels
site <- rep(growth_data$Site, 2)

# Create corresponding Year labels
year <- c(rep("2005", nrow(growth_data)), rep("2020", nrow(growth_data)))

# Combine Site and Year into one factor
site_year <- interaction(site, year)

# Make nicer labels
levels(site_year) <- c("Northeast 2005", "Southwest 2005",
                       "Northeast 2020", "Southwest 2020")

# Box plot
boxplot(circumference ~ site_year,
        main = "Tree Circumference at Start (2005) and End (2020)",
        xlab = "Site and Year",
        ylab = "Circumference (cm)",
        col = c("goldenrod", "cornflowerblue", "goldenrod", "cornflowerblue"))
```

#### 9. Calculate the mean growth over the last 10 years at each site.
The mean growth over the last 10 years at each site is:

* Northeast site: 42.94 cm
* Southwest site: 35.49 cm

```{r}
# Calculate 10-year growth per tree using 2020 and 2010
growth_data$growth10yr <- growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm

# Mean growth per site
growth_summary <- aggregate(growth10yr ~ Site, data = growth_data, mean)

#Print out growth summary
growth_summary 
```

We use `diff` on the `growth10yr` column to subtract the values. So, on average, trees at the northeast site grew *~7.45cm* more over the last 10 years than trees at the southwest site.
```{r}
# subtract the growth10yr values from each other
diff(growth_summary$growth10yr)
```

#### 10. Use the t.test to estimate the p-value that the 10 year growth is different at the two sites.
A Welch two-sample t-test indicated that this difference was not statistically significant (t = 1.8882, df = 87.978, p = 0.06229), with a 95% confidence interval of -0.39 to 15.29 cm. The p-value is slightly above the common significance threshold of 0.05.

```{r}
t.test(growth10yr ~ Site, data = growth_data)
```

## Part 2: Examining biological sequence diversity

##### Organisms for comparison
In this section we are studying Acetobacter aceti (GCA_002005445) and comparing its sequence features to E. coli (GCA_000005845).

* Acetobacter aceti: gram-negative bacterium known for production of acetic acid and used for converting ethanol into vinegar (acetic acid).
* Escherichia coli: gram-negative bacterium. E. coli can ferment glucose to produce acetate, but this growth is inhibited by high acetate concentrations. It plays a role in the gut microbiota and in food production, but is also a common pathogen, causing a range of diseases. 

#### 1. Download the whole set of coding DNA sequences for E. coli and your organism of interest. How many coding sequences are present in these organisms? Present this in the form of a table. Describe any differences between the two organisms. 
Download both E. coli (GCA_000005845) and A. aceti (GCA_002005445) from Ensembl Bacteria.

```{r}
#E. coli
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")
gunzip("ecoli_cds.fa.gz")

#A. aceti
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_80_collection/acetobacter_aceti_gca_002005445/cds/Acetobacter_aceti_gca_002005445.ASM200544v1.cds.all.fa.gz"
download.file(URL,destfile="aaceti_cds.fa.gz")
gunzip("aaceti_cds.fa.gz")
list.files()
```

We can figure out the number of coding sequences (CDS) and present this in a table called `cds_table`.
As below, E. coli has around **837** more coding sequences than A. aceti (4239 - 3402 = 837).
```{r}
ecoli_cds <- seqinr::read.fasta("ecoli_cds.fa")
aaceti_cds <- seqinr::read.fasta("aaceti_cds.fa")

cds_table <- data.frame(
  Organism = c("E. coli", "A. aceti"),
  CDS_Count = c(length(ecoli_cds), length(aaceti_cds))
)

# Print table
cds_table

# Calculate difference in CDS count
diff(cds_table$CDS_Count)
```

#### 2. How much coding DNA is there in total for these two organisms? Present this in the form of a table. Describe any differences between the two organisms.
The total coding DNA in E. coli is about **741,960** base pairs greater than in A. aceti. This difference is consistent with its larger number of genes as noted in question 1.

We want all rows and first column in the cds files to be summed. This will be the total length of all genes.
```{r}
# Total coding DNA from the length column
ecoli_len <- as.numeric(summary(ecoli_cds)[,1])
aaceti_len <- as.numeric(summary(aaceti_cds)[,1])

# Sum of all the lengths within the organism 
ecoli_total <- sum(ecoli_len)
aaceti_total <- sum(aaceti_len)

# Create table
DNA_total <- data.frame(
  Organism = c("E. coli", "A. aceti"),
  Total_DNA = c(ecoli_total, aaceti_total)
)

# Print table
DNA_total

# Difference in Total_DNA
diff(DNA_total$Total_DNA)
```

#### 3. Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms. What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms.

On average, A.aceti coding sequences are longer than those in E. coli (approximately 12 bp difference) but this is very small relative to the overall sequence length and might not be biologically significant.

The median coding sequence length is longer in E. coli (831 bp vs 819 bp), which suggests that E. coli has a slightly more balanced distribution of gene lengths, whereas A. aceti may have a few longer genes that increase its mean.

The fact that A. aceti has a higher mean but a lower median implies its coding sequence length distribution might be more skewed, with some unusually long genes pulling the mean upward.
```{r}
# Boxplot of CDS lengths - check if this is ecoli_len or is it the sum (ecoli_total)
boxplot(list("E. coli" = ecoli_len, "A. aceti" = aaceti_len),
        main = "Distribution of coding sequence lengths (box plot)",
        ylab = "Sequence Length (bp)",
        col = c("skyblue", "plum1"))

# Calculate mean and median CDS length
cds_stats <- data.frame(
  Organism = c("E. coli", "A. aceti"),
  Mean_CDS_Length = c(mean(ecoli_len), mean(aaceti_len)),
  Median_CDS_Length = c(median(ecoli_len), median(aaceti_len))
)

# Print table to show means and medians
cds_stats

#Calculate differences in mean and median coding sequences
diff(cds_stats$Mean_CDS_Length)
diff(cds_stats$Median_CDS_Length)
```

#### 4. Calculate the frequency of DNA bases in the total coding sequences for both organisms. Perform the same calculation for the total protein sequence. Create bar plots for nucleotide and amino acid frequency. Describe any differences between the two organisms.

##### Calculate the frequency of DNA bases in the total coding sequences for both organisms.
To count the entire genome we use `unlist`.
```{r}
# Concatenate all CDS into one sequence per organism
ecoli_seq <- unlist(ecoli_cds)
aaceti_seq <- unlist(aaceti_cds)

# Count combinations of 1s of the nucleotides in entire DNA using seqinr::count
ecoli_dna_composition <- seqinr::count(ecoli_seq,1)
aaceti_dna_composition <- seqinr::count(aaceti_seq,1)

#Print nucleotide composition
ecoli_dna_composition
aaceti_dna_composition

# Relative frequencies can also be calculated using freq = TRUE
ecoli_nt_freq <- seqinr::count(ecoli_seq,1, freq = TRUE)
aaceti_nt_freq <- seqinr::count(aaceti_seq,1, freq = TRUE)

#Print relative frequencies
ecoli_nt_freq 
aaceti_nt_freq
```
##### Create bar plots for nucleotide frequency.
```{r}
#Barplots for our nucleotide frequencies
barplot(ecoli_nt_freq , 
        xlab="Nucleotides", 
        ylab="Proportion", 
        main="E. coli CDS composition", 
        las=1, # to make the labels horizontal only
        col="skyblue")
grid()
barplot(aaceti_nt_freq, 
        xlab="Nucleotides", 
        ylab="Proportion", 
        main="A. aceti CDS composition", 
        las=1, # to make the labels horizontal only
        col="plum1")
grid()
```


##### Translation for the total protein sequence of the genes
In the code, `lapply` applies the translate function to each element of the cds list, resulting in prot, a list of translated protein sequences.
```{r}
#unlist and use lapply to translate entire genome
ecoli_prot <- unlist(lapply(ecoli_cds, translate))
aaceti_prot <- unlist(lapply(aaceti_cds, translate))
```

We can use the two prot lists to get any useful information, such as the number of unique amino acids.
Here we store E. coli in `aa` and A. aceti in `aa1`. Both have 20 unique amino acids.
```{r}
# Find unique number of amino acids in E. coli
aa <- unique(ecoli_prot)
aa <- aa[aa != "*"] # removes any asterisks for unknown amino acids
length(aa) # counts the remaining unique amino acids

# Find unique number of amino acids in A. aceti
aa1 <- unique(aaceti_prot)
aa1 <- aa1[aa1 != "*"]
length(aa1) # counts the remaining unique amino acids
```

We can use `aa` and `aa1` to then determine the count of each amino acid coded by E. coli and A. aceti, respectively. The following code counts the occurence of each amino acid using the total set of possible amino acids.
```{r}
# Count amino acids using seqinr::count
ecoli_aa_count <- seqinr::count(ecoli_prot,wordsize=1,alphabet=aa)
aaceti_aa_count <- seqinr::count(aaceti_prot,wordsize=1,alphabet=aa1)

# Print total amino acid count
ecoli_aa_count
aaceti_aa_count
```

##### Create bar plots for amino acid count
The barplots show the frequency of protein sequences for both organisms as shown in `ecoli_aa_count` and `aaceti_aa_count`.
```{r}
# Protein amino acid frequencies
barplot(ecoli_aa_count,
        main="E. coli Amino Acid Count",
        ylab="Count", 
        col="skyblue", 
        las=1, # keep labels horizontal
        names.arg = names(ecoli_aa_count)) # add AA labels
grid()

barplot(aaceti_aa_count,
        main = "A. aceti Amino Acid Count",
        ylab = "Count",
        col = "plum1",
        las = 1,
        names.arg = names(aaceti_aa_count))
grid()
```
We can then look at the barplots for the frequencies together using `ggplot2` and do a side-by-side comparison of the two organisms' amino acid composition. While both organisms display broadly similar distributions, we can see A. aceti shows higher proportions of alanine (A), glycine (G), proline (P), arginine (R), and serine (S), whereas E. coli is enriched in isoleucine (I), lysine (K), asparagine (N), and glutamine (Q).
```{r}
# Convert to proportions to view together
ecoli_aa_proportion <- seqinr::count(ecoli_prot,wordsize=1,alphabet=aa, freq = TRUE)
aaceti_aa_proportion <- seqinr::count(aaceti_prot,wordsize=1,alphabet=aa1, freq = TRUE)


# Merge into one data frame (long format for ggplot2)
aa_df <- data.frame(
  AA = names(ecoli_aa_proportion),
  E_coli = as.numeric(ecoli_aa_proportion),
  A_aceti = as.numeric(aaceti_aa_proportion)
)

# Reshape into long format
aa_long <- reshape2::melt(aa_df, id.vars = "AA",
                          variable.name = "Organism",
                          value.name = "Proportion")

# Plot side-by-side comparison
ggplot(aa_long, aes(x = AA, y = Proportion, fill = Organism)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  labs(title = "Amino Acid Proportion Comparison",
       x = "Amino Acid",
       y = "Proportion") +
  scale_fill_manual(values = c("skyblue", "plum1")) +
  theme_minimal(base_size = 14)
```

We can also generate a table to compare the two organisms' nucleotides and amino acid sequences side-by-side.
Here `knitr` is used to generate a cleaner html table.
```{r}
# DNA nucleotide frequency table
nt_freq_table <- data.frame(
  Base = union(names(ecoli_nt_freq), names(aaceti_nt_freq)),
  E_coli = as.numeric(ecoli_nt_freq[union(names(ecoli_nt_freq), names(aaceti_nt_freq))]),
  A_aceti = as.numeric(aaceti_nt_freq[union(names(ecoli_nt_freq), names(aaceti_nt_freq))])
)

# Replace NAs with 0
nt_freq_table[is.na(nt_freq_table)] <- 0

knitr::kable(nt_freq_table, digits=4,
             col.names = c("Base", "E. coli (Proportion)", "A. aceti (Proportion)"))

# Amino acid frequency table
aa_freq_table <- data.frame(
  AminoAcid = union(names(ecoli_aa_proportion), names(aaceti_aa_proportion)),
  E_coli = as.numeric(ecoli_aa_proportion[union(names(ecoli_aa_proportion), names(aaceti_aa_proportion))]),
  A_aceti = as.numeric(aaceti_aa_proportion[union(names(ecoli_aa_proportion), names(aaceti_aa_proportion))])
)

# Replace NAs with 0
aa_freq_table[is.na(aa_freq_table)] <- 0

#generate table using knitr (cleaner if viewing in html)
knitr::kable(aa_freq_table, digits=4,
             col.names = c("Amino Acid", "E. coli (Proportion)", "A. aceti (Proportion)"))
```

##### Describe any differences between the two organisms.

**Nucleotides**

From our results regarding the frequency of DNA bases, we can see that E. coli and A. aceti differ noticeably in base composition.

* E. coli is more balanced between A/T (both ~24%) and G/C (~25–27%).
* A. aceti is relatively GC-rich, with higher C (28%) and G (29.6%) but lower A (20%) and T (21.8%).
* This indicates that A. aceti has a higher GC content compared to E. coli, which could potentially influence genome stability, codon usage (explored later) and gene expression.

**Amino Acid**

* Both E. coli and A. aceti show broadly similar amino acid usage.
* Both organisms have Leucine (L) as the most abundant amino acid (~10%) followed by Alanine (A) and Glycine (G).
* Cysteine (C) and Tryptophan (W) are rare in both (<2%).
* Differences are A. aceti shows higher proportions of alanine (A: 0.1098 vs 0.0954), glycine (G: 0.0831 vs 0.0736), proline (P: 0.0525 vs 0.0444), arginine (R: 0.0692 vs 0.0553), and serine (S: 0.0642 vs 0.0578).
* In contrast, E. coli is enriched in isoleucine (I: 0.0601 vs 0.0524), lysine (K: 0.0440 vs 0.0325), asparagine (N: 0.0390 vs 0.0287), and glutamine (Q: 0.0445 vs 0.0332). These differences likely reflect adaptations to the metabolic and environmental requirements of each organism.

#### 5. Create a codon usage table and quantify the codon usage bias among all coding sequences. Describe any differences between the two organisms with respect to their codon usage bias. Provide charts to support your observations.

##### Create a codon usage table and quantify the codon usage bias among all coding sequences
A codon is a three base triplet that codes for a particular amino acid. As there are 64 possible codons and just 20 amino acids (and 3 termination codons); some amino acids are coded by multiple codons. For example, leucine is coded by six codons that are synonymous for it.

`seqinr` has a command called `uco` which calculates some codon usage information.

The RSCU stands for Relative Synonymous Codon Usage, which is a measure as to how preferred a codon of a particular amino acid is compared to the other codons which also encode it.

* RSCU ≈ 1 means codon used as expected.
* RSCU > 1 means codon is preferred.
* RSCU < 1 means codon is under-represented.

We create a codon usage table to quantify the codon usage bias for both E. coli and A. aceti. The first 6 rows are displayed.
```{r}
# Codon usage tables
# Concatenate all CDS into one long sequence and get codon frequencies
ecoli_codon <- uco(unlist(ecoli_cds),index="rscu", as.data.frame = TRUE)
aaceti_codon <- uco(unlist(aaceti_cds), index = "rscu", as.data.frame = TRUE)

#View top 6 rows of both tables
head(ecoli_codon)
head(aaceti_codon)
```

#####Provide charts to support your observations.

We can display bar plots to visualise codon usage for both organisms. The `las=2` is to keep the labels vertical for ease of graph reading.The red dashed line at RSCU = 1 shows where codons would be equally used.
```{r}
#store as non-dataframe for below graphs
ecoli_rscu <- uco(unlist(ecoli_cds),index="rscu")
aaceti_rscu <- uco(unlist(aaceti_cds), index = "rscu")

barplot(ecoli_rscu, las=2, cex.names=0.5,
        main="E. coli RSCU", ylab="RSCU", col="skyblue")
abline(h=1, col="red", lty=2)

barplot(aaceti_rscu, las=2, cex.names=0.5,
        main="A. aceti RSCU", ylab="RSCU", col="plum1")
abline(h=1, col="red", lty=2)
```

However, it is better if we could look at the RSCU of both organisms side by side. Therefore, we can create a new combined table.
We can combine `ecoli_codon` and `aaceti_codon` into one table with the amino acid and codon columns and call it `rscu_table`.
```{r}
# Make sure RSCU columns have unique names
colnames(ecoli_codon)[colnames(ecoli_codon) == "RSCU"] <- "RSCU_Ecoli"
colnames(aaceti_codon)[colnames(aaceti_codon) == "RSCU"] <- "RSCU_Aaceti"

# Merge by Codon (you can keep either AA column; they should match)
rscu_table <- merge(ecoli_codon[, c("codon", "AA", "RSCU_Ecoli")],
                    aaceti_codon[, c("codon", "RSCU_Aaceti")],
                    by = "codon")

# Reorder columns
rscu_table <- rscu_table[, c("codon", "AA", "RSCU_Ecoli", "RSCU_Aaceti")]

# print top 6 rows table
head(rscu_table)
```

To further assist with visualisation of codon usage bias in both organisms, the below creates a heatmap for codon bias using `ggplot2`. Here the codons are aligned on the x-axis, with the organisms in the y-axis. We can compare the RSCU visually this way.
```{r}
# Convert to data frames to prepare for heat map (different to the rscu_table)
df_ecoli <- data.frame(
  Codon = names(ecoli_rscu),
  RSCU = as.numeric(ecoli_rscu),
  Organism = "E. coli"
)

df_aaceti <- data.frame(
  Codon = names(aaceti_rscu),
  RSCU = as.numeric(aaceti_rscu),
  Organism = "A. aceti"
)

# Combine
rscu_all <- rbind(df_ecoli, df_aaceti)

# Plot heat map with ggplot2
ggplot(rscu_all, aes(x = Codon, y = Organism, fill = RSCU)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(low = "blue4", mid = "white", high = "deeppink4",
                       midpoint = 1, name = "RSCU") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "Codon Usage Bias Heatmap (RSCU values)",
       x = "Codon", y = "Organism")

```

If we group the codons by the amino acids we can further visualise the most common ones with the codon bias for each. The stacked bar graph is clearer in showing the contribution of the codons to a specific amino acid.

This stacked bar plot uses `ggplot2`, `dplyr`, `patchwork` and `forcats` (all called earlier in this R Markdown).
Note that we have 21 amino acids as **Stp** is the extra amino acid (stop codon) coded by TAA, TAG and TGA.

The below are relative synonymous codon usage (RSCU) charts for E. coli and A. aceti where all codons coding for each amino acid are represented in the boxes below the bar chart.
```{r}
# Prepare data for plot
dat_aaceti <- data.frame(Codon=c(rscu_table$codon),RSCU=c(rscu_table$RSCU_Aaceti),AA=c(rscu_table$AA))
dat_ecoli <- data.frame(Codon=c(rscu_table$codon),RSCU=c(rscu_table$RSCU_Ecoli),AA=c(rscu_table$AA))

# Create a column for fill colour of each codon
dat_aaceti$Col=1
dat_ecoli$Col=1

# This populates the Col column with unique values for each codon coding an amino. Each value is a colour, therefore all amino acids will share the same values but the codons will be different.
dat_aaceti=dat_aaceti %>% group_by(AA) %>% mutate(Col = lag(cumsum(Col), default = 0)) %>% mutate(Col =as.factor(Col))
dat_ecoli=dat_ecoli %>% group_by(AA) %>% mutate(Col = lag(cumsum(Col), default = 0)) %>% mutate(Col =as.factor(Col))

# Create bar plots - A. aceti
p0 = ggplot(data=dat_aaceti, aes(x=AA, y=RSCU, fill=Col)) +
  geom_bar(stat='identity') +
  theme_minimal() +
  theme(axis.title.x = element_blank()) +
  guides(fill = "none") +
  ggtitle("Codon Usage Bias in A. aceti") +
  theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))

# Create tile plot - A. aceti
p1 = ggplot(data=dat_aaceti) +
  geom_tile(aes(x=AA, y=fct_rev(Col), fill=Col), col="white") +
  geom_text(aes(x=AA, y=Col, label=Codon), col="white", size=3.5) +
  theme_void() +
  theme(legend.position="none") +
  guides(fill = "none")

# Create bar plots - E. coli
p2 = ggplot(data=dat_ecoli, aes(x=AA, y=RSCU, fill=Col)) +
  geom_bar(stat='identity') +
  theme_minimal() +
  theme(axis.title.x = element_blank()) +
  guides(fill = "none") +
  ggtitle("Codon Usage Bias in E. coli") +
  theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))

# Create tile plot - E. coli
p3 = ggplot(data=dat_ecoli) +
  geom_tile(aes(x=AA, y=fct_rev(Col), fill=Col), col="white") +
  geom_text(aes(x=AA, y=Col, label=Codon), col="white", size=3.5) +
  theme_void() +
  theme(legend.position="none") +
  guides(fill = "none")

#Combine plots in a single plot for each organism
plot_aaceti=p0/p1
plot_ecoli=p2/p3

#Set plot size. The ratio here is 7:1 - i.e. the bar plot is 7 times the height of the tile plot
plot_aaceti=plot_aaceti+plot_layout(heights = c(7, 1))
plot_ecoli=plot_ecoli+plot_layout(heights = c(7, 1))

#print plots
plot_ecoli
plot_aaceti
```

We can further separate each amino acid into distinct plots. The "Stp" amino acid is removed for clarity in this plot.

A. aceti plot:
```{r}
## Store data into a new data frame structure
aaceti_dat <- data.frame(
  Codon = rscu_table$codon,
  RSCU  = rscu_table$RSCU_Aaceti,
  AA    = rscu_table$AA,
  stringsAsFactors = FALSE
)

# Remove stop codons
aaceti_dat <- aaceti_dat[aaceti_dat$AA != "Stp", ]

# Order by AA then Codon
aaceti_dat <- aaceti_dat[order(aaceti_dat$AA, aaceti_dat$Codon), ]
aa_levels <- unique(aaceti_dat$AA)

# Set plotting area: adjust nrow/ncol depending on number of amino acids
nAA <- length(aa_levels)
ncol <- 5   # choose grid layout
nrow <- ceiling(nAA / ncol)

par(mfrow=c(nrow, ncol))
par(mar=c(3,3,2,1), oma=c(3,1,3,1), cex.axis=0.8, cex.lab=0.8, cex.main=0.9)

# Loop through each AA and plot codon usage
for (current_aa in aa_levels) {
  subset_aaceti_dat <- aaceti_dat[aaceti_dat$AA == current_aa, ]
  
  bp <- barplot(
    subset_aaceti_dat$RSCU,
    names.arg = subset_aaceti_dat$Codon,
    ylim = c(0, max(aaceti_dat$RSCU) + 0.5),
    col = "plum1",
    las = 2,
    cex.names = 0.8,
    ylab = "RSCU",
    main = current_aa
  )
  
  # Reference line at RSCU = 1
  abline(h=1, col="red", lty=2)
}

# Add a main title for the entire figure
mtext("Codon Usage Bias Across Amino Acids (A. aceti)", outer = TRUE, cex = 0.9, font = 2)
```

For E. coli:
```{r}
## Store data into a new data frame structure
ecoli_dat <- data.frame(
  Codon = rscu_table$codon,
  RSCU  = rscu_table$RSCU_Ecoli,
  AA    = rscu_table$AA,
  stringsAsFactors = FALSE
)

# Remove stop codons
ecoli_dat <- ecoli_dat[ecoli_dat$AA != "Stp", ]

# Order by AA then Codon
ecoli_dat <- ecoli_dat[order(ecoli_dat$AA, ecoli_dat$Codon), ]
aa_levels_1 <- unique(ecoli_dat$AA)

# Set plotting area: adjust nrow/ncol depending on number of amino acids
nAA <- length(aa_levels_1)
ncol <- 5   # choose grid layout
nrow <- ceiling(nAA / ncol)

par(mfrow=c(nrow, ncol))
par(mar=c(3,3,2,1), oma=c(3,1,3,1), cex.axis=0.8, cex.lab=0.8, cex.main=0.9)

# Loop through each AA and plot codon usage
for (current_aa_1 in aa_levels_1) {
  subset_ecoli_dat <- ecoli_dat[ecoli_dat$AA == current_aa_1, ]
  
  bp1 <- barplot(
    subset_ecoli_dat$RSCU,
    names.arg = subset_ecoli_dat$Codon,
    ylim = c(0, max(ecoli_dat$RSCU) + 0.5),
    col = "skyblue",
    las = 2,
    cex.names = 0.8,
    ylab = "RSCU",
    main = current_aa_1
  )
  
  # Reference line at RSCU = 1
  abline(h=1, col="red", lty=2)
}

# Add a main title for the entire figure
mtext("Codon Usage Bias Across Amino Acids (E. coli)", outer = TRUE, cex = 0.9, font = 2)
```
##### Describe any differences between the two organisms with respect to their codon usage bias. 
Both organisms exhibit codon usage bias, with some codons for the same amino acid used more frequently in one organism than the other

* The CTG codon (codes for Leu) is most overrepresented in both organisms as shown in the heatmap and the codon usage stacked bargraph
* Analysis of RSCU values shows that E. coli exhibits strong codon usage bias, favoring specific codons for several amino acids (e.g., AAA for Lys, CCG for Pro, CTG for Leu) while underusing others.
* A. aceti displays a more balanced codon usage, for example, AAG is slightly preferred over AAA for Lys, CCC over CCG for Pro, and TGA is the dominant stop codon instead of TAA.
* Other amino acids, such as Gly and Ala, show similar patterns in both organisms.
* E. coli tends to have more extreme biases as clearly shown on our heatmap.
* The differences may reflect organism-specific adaptations in protein structure, metabolic needs and protein translational efficiency. Our side-by-side barplots of amino acid proportions highlight these subtle but consistent shifts between the two species.

#### 6. In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most over- and under-represented k-mers in your organism of interest. Are these k-mers also over- and under-represented in E. coli to a similar extent? Provide plots to support your observations. Why do you think these sequences are present at different levels in the genomes of these organisms?

Before, we already translated the CDS to protein sequences. K-mer is simply a short sequence string of length k. We can use k-mer analysis to identify particular protein sequences that are over- or under-represented in an organism.

##### 3-kmers
We counted unique amino acids in ecoli earlier and stored as `aa` (E. coli) and `aa1` (A. aceti).
```{r}
#Prepare file
ecoli_kmer <- unlist(ecoli_prot)
aaceti_kmer <- unlist(aaceti_prot)

#For 3 kmers
ecoli_k3 <- seqinr::count(ecoli_kmer,wordsize=3,alphabet=aa)
aaceti_k3 <- seqinr::count(aaceti_kmer,wordsize=3,alphabet=aa1)
#4 k mers
ecoli_k4 <- seqinr::count(ecoli_kmer,wordsize=4,alphabet=aa)
aaceti_k4 <- seqinr::count(aaceti_kmer,wordsize=4,alphabet=aa1)
#5 kmers
ecoli_k5 <- seqinr::count(ecoli_kmer,wordsize=5,alphabet=aa)
aaceti_k5 <- seqinr::count(aaceti_kmer,wordsize=5,alphabet=aa1)

#check all
head(ecoli_k3)
head(aaceti_k3)

head(ecoli_k4)
head(aaceti_k4)

head(ecoli_k5)
head(aaceti_k5)

#Convert counts to frequency so the different proteome sizes don't bias the comparison
ecoli_k3_freq <- seqinr::count(ecoli_kmer,wordsize=3,alphabet=aa,freq=TRUE)
head(ecoli_k3_freq)
aaceti_k3_freq <- seqinr::count(aaceti_kmer,wordsize=3,alphabet=aa1,freq=TRUE)
head(aaceti_k3_freq)

ecoli_k4_freq <- seqinr::count(ecoli_kmer,wordsize=4,alphabet=aa,freq=TRUE)
head(ecoli_k4_freq)
aaceti_k4_freq <- seqinr::count(aaceti_kmer,wordsize=4,alphabet=aa1,freq=TRUE)
head(aaceti_k4_freq)

ecoli_k5_freq <- seqinr::count(ecoli_kmer,wordsize=5,alphabet=aa,freq=TRUE)
head(ecoli_k5_freq)
aaceti_k5_freq <- seqinr::count(aaceti_kmer,wordsize=5,alphabet=aa1,freq=TRUE)
head(aaceti_k5_freq)
```

Place frequencies into a table together.
```{r}
k3_table <- data.frame(
    Ecoli = ecoli_k3_freq,
  Aaceti = aaceti_k3_freq
)

k4_table <- data.frame(
    Ecoli = ecoli_k4_freq,
  Aaceti = aaceti_k4_freq
)

k5_table <- data.frame(
    Ecoli = ecoli_k5_freq,
  Aaceti = aaceti_k5_freq
)

#view top 6
head(k3_table)
head(k4_table)
head(k5_table)
```
Finding the top 10 over and under represented k-mers in A. aceti for length 3-5.

This plots the bar graph with both top and bottom.

Still yet to: "Are these k-mers also over- and under-represented in E. coli to a similar extent?"
We could pull out the same kmers as the k3_table would have e. coli 

This selects the 10 highest and 10 lowest frequency kmers in A. aceti.
```{r}
#selects the 10 highest and 10 lowest frequency kmers in A. aceti for k-mer of 3
over_k3_aaceti <- head(k3_table[order(-k3_table$Aaceti.Freq), ], 10)
under_k3_aaceti <- head(k3_table[order(k3_table$Aaceti.Freq), ], 10)
aaceti_top_k3 <- rbind(over_k3_aaceti, under_k3_aaceti)

#selects the 10 highest and 10 lowest frequency kmers in A. aceti for k-mer of 4
over_k4_aaceti <- head(k4_table[order(-k4_table$Aaceti.Freq), ], 10)
under_k4_aaceti <- head(k4_table[order(k4_table$Aaceti.Freq), ], 10)
aaceti_top_k4 <- rbind(over_k4_aaceti, under_k4_aaceti)

#selects the 10 highest and 10 lowest frequency kmers in A. aceti for k-mer of 5
over_k5_aaceti <- head(k5_table[order(-k5_table$Aaceti.Freq), ], 10)
under_k5_aaceti <- head(k5_table[order(k5_table$Aaceti.Freq), ], 10)
aaceti_top_k5 <- rbind(over_k5_aaceti, under_k5_aaceti)
```

For the top 10 highest and 10 lowest frequency kmers in A. aceti, we compare the frequencies with E. coli - i.e. for the same k-mer motif. We will use a function to graphs for each k-mer (of length 3-5).

What the below code chunk does is:

* Pull out the 10 most frequent and 10 least frequent k-mers in A. aceti.
* Show how frequent (or rare) those same k-mers are in E. coli.
* This way we directly see whether the enrichment/depletion is specific to A. aceti or shared with E. coli.

The below graphs display:

* One for the top 10 (most over-represented)
* Another for the bottom 10 (most under-represented)
```{r}
plot_kmer_comparison_split <- function(prot1, prot2, aa1, aa2, k = 3, top_n = 10,
                                       org1_name="A. aceti", org2_name="E. coli") {
  
  # Count k-mers (frequency)
  kmers1_freq <- seqinr::count(prot1, wordsize = k, alphabet = aa1, freq = TRUE)
  kmers2_freq <- seqinr::count(prot2, wordsize = k, alphabet = aa2, freq = TRUE)
  
  # Convert to data frame
  kmer_table <- data.frame(
    Kmer = names(kmers1_freq),
    Aaceti.Freq = as.numeric(kmers1_freq),
    Ecoli.Freq = as.numeric(kmers2_freq),
    stringsAsFactors = FALSE
  )
  
  # Select top and bottom k-mers in A. aceti
  over_kmers  <- head(kmer_table[order(-kmer_table$Aaceti.Freq), ], top_n)
  under_kmers <- head(kmer_table[order(kmer_table$Aaceti.Freq), ], top_n)
  
  # Overrepresented plot
  freq_matrix_over <- t(as.matrix(over_kmers[, c("Aaceti.Freq", "Ecoli.Freq")]))
  colnames(freq_matrix_over) <- over_kmers$Kmer
  
  barplot(
    freq_matrix_over,
    beside = TRUE,
    col = c("plum1", "skyblue"),
    las = 2,
    ylab = "Frequency",
    xlab = paste(k, "-mers", sep=""),
    main = paste("Top", top_n, "Over-represented", k, "-mers in", org1_name, "vs", org2_name),
    cex.names = 0.8
  )
  legend("topright", legend = c(org1_name, org2_name), fill = c("plum1", "skyblue"))
  
  # Underrepresented plot
  freq_matrix_under <- t(as.matrix(under_kmers[, c("Aaceti.Freq", "Ecoli.Freq")]))
  colnames(freq_matrix_under) <- under_kmers$Kmer
  
  barplot(
    freq_matrix_under,
    beside = TRUE,
    col = c("plum1", "skyblue"),
    las = 2,
    ylab = "Frequency",
    xlab = paste(k, "-mers", sep=""),
    main = paste("Top", top_n, "Under-represented", k, "-mers in", org1_name, "vs", org2_name),
    cex.names = 0.8
  )
  legend("topright", legend = c(org1_name, org2_name), fill = c("plum1", "skyblue"))
}

# Call function for each kmer = 3
plot_kmer_comparison_split(aaceti_prot, ecoli_prot, aa1, aa, k=3)
# Call function for each kmer = 4
plot_kmer_comparison_split(aaceti_prot, ecoli_prot, aa1, aa, k=4)
# Call function for each kmer = 5
plot_kmer_comparison_split(aaceti_prot, ecoli_prot, aa1, aa, k=5)
```

We can quantify the changes using a table and use a function for efficiency across 3-5 kmers.
```{r}
#create function
get_kmer_summary <- function(prot1, prot2, aa1, aa2, k = 3, top_n = 10,
                             org1_name="A. aceti", org2_name="E. coli") {
  
  # Count k-mers (frequency)
  kmers1_freq <- seqinr::count(prot1, wordsize = k, alphabet = aa1, freq = TRUE)
  kmers2_freq <- seqinr::count(prot2, wordsize = k, alphabet = aa2, freq = TRUE)
  
  # Convert to data frame
  kmer_table <- data.frame(
    Kmer = names(kmers1_freq),
    Aaceti.Freq = as.numeric(kmers1_freq),
    Ecoli.Freq = as.numeric(kmers2_freq),
    stringsAsFactors = FALSE
  )
  
  # Add fold change (handle zeros)
  kmer_table$FoldChange <- ifelse(kmer_table$Ecoli.Freq > 0,
                                  kmer_table$Aaceti.Freq / kmer_table$Ecoli.Freq,
                                  NA)
  
  # Select over- and under-represented k-mers in Org1
  over_kmers  <- head(kmer_table[order(-kmer_table$Aaceti.Freq), ], top_n)
  under_kmers <- head(kmer_table[order(kmer_table$Aaceti.Freq), ], top_n)
  
  # Combine into one summary table
  summary_table <- rbind(
    cbind(Type="Over-represented", over_kmers),
    cbind(Type="Under-represented", under_kmers)
  )
  
  return(summary_table)
}

#use function and print
summary_3kmers <- get_kmer_summary(aaceti_prot, ecoli_prot, aa1, aa, k=3)
summary_4kmers <- get_kmer_summary(aaceti_prot, ecoli_prot, aa1, aa, k=4)
summary_5kmers <- get_kmer_summary(aaceti_prot, ecoli_prot, aa1, aa, k=5)
print(summary_3kmers)
print(summary_4kmers)
print(summary_5kmers)
```

##### Are these k-mers also over- and under-represented in E. coli to a similar extent?
We analysed the protein sequences of A. aceti and compared them to E. coli to identify k-mers (length 3–5 amino acids) that are significantly over- or under-represented in A. aceti. Frequencies were calculated relative to the total proteome to account for differences in proteome size.

* **3-mer**
  * The ten most over-represented 3-mers in A. aceti include AAL, AAA, ALA, LAA, ALL, LLA, AGL, AAG, LAL, and LAG. 
  * Most of these k-mers are also present in E. coli but at slightly lower frequencies, indicating a species-specific enrichment. * Under-represented 3-mers include CMY, CQC, KCC, MCY, WKC, WWC, YCK, CCC, CCM, and CHW, which remain rare or absent in both organisms.

* **4-mer**
  * Over-represented 4-mers in A. aceti include AALA, ALAA, AAAA, LAAL, ALLA, AAAL, LAAA, LLAA, AGLA, and ALAL. 
  * Many of these 4-mers show modest representation in E. coli, but the enrichment in A. aceti suggests organism-specific sequence preferences.
  * Under-represented 4-mers are mostly absent in A. aceti, while E. coli has very low but nonzero frequencies.
  
* **5-mer**
  * The top 5-mers enriched in A. aceti include GKSTL, AALAA, VSSGG, AAAAL, AAALA, AAALG, ALAAG, LAAAL, LAALR, and AARAA.
  * Many of these sequences are either absent or extremely rare in E. coli, highlighting strong organism-specific over-representation.
  * Under-represented 5-mers are largely absent in A. aceti and present only at trace levels in E. coli.
  
##### Why do you think these sequences are present at different levels in the genomes of these organisms?

Some factors influencing k-mer frequencies could be:

  * **GC content:** the genome base composition as explored in questions 1-4 showed there could be different GC content which in turn can affect amino acid usage and recurring/repeated peptides
  * **Codon usage and amino acid frequencies:** organism specific codon bias as shown in question 5 can lead to more of certain amino acids and therefore their recurring short motifs (k-mers)
  * **Ecological niche:** adaptations to different environments may favor proteins with structural motifs that are more stable or functional under niche-specific conditions. E. coli lives in the gut, so its proteins might favor motifs aiding stress resistance and rapid growth. A. aceti likes acidic environments, possibly selecting for motifs stabilising proteins in low pH.
  * **Mutations and evolution over time:** over time, sequence composition may change due to selection and drift which could lead to some k-mers being amplified over others.

In summary, these factors explain why A. aceti shows enrichment of k-mers like AAL, AAA, and AALA, while other k-mers remain rare or absent compared to E. coli. The differences in k-mer presence and frequency could reflect the unique genetic histories and functional adaptations of E. coli and A. aceti which are shaped by factors like their environment, genetic content and evolution.