---
title: "Assessment 4: Bioinformatics R Project"
author: "Julia Lay"
date: "2025-09-28"
output: 
  html_document:
    df_print: kable
    toc: true
    toc_depth: 4
    toc_float: true
bibliography: "references.bib"
csl: "harvard-deakin-university.csl"
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## A. Setup: Install packages if required

```{r install packages}
# Function to check if packages are installed
# Install missing packages quietly
install_if_missing <- function(pkgs) {
  for (pkg in pkgs) {
    if (!requireNamespace(pkg, quietly = TRUE)) {
      message("Installing package: ", pkg)
      suppressMessages(
        suppressWarnings(
          install.packages(pkg, dependencies = TRUE, quiet = TRUE)
        )
      )
    }
  }
}

# List of required packages
my_packages <- c(
  "seqinr", # sequence data processing
  "R.utils", # utilities (gunzip, etc.)
  "ggplot2", # plotting
  "dplyr", # data wrangling
  "patchwork", # combine plots
  "forcats", # factor manipulation (fct_rev())
  "knitr" # knitting and html tables
)

# Install any missing packages
install_if_missing(my_packages)
```


## B. Setup: Load all required packages

Note that the `count()` function must be distinguished within code chunks as it exists in both `seqinr` and `dplyr`.

```{r loading packages}
suppressPackageStartupMessages({
  library("seqinr")
  library("R.utils")
  library("ggplot2")
  library("dplyr")
  library("patchwork")
  library("forcats")
  library("knitr")
})
```

## 1. Gene Expression and Growth Data Analysis

### Gene Expression Analysis

#### 1.1. Read in the file and show table with first six genes
The below reads in the file `gene_expression.tsv` making the gene identifiers the row names. The `head()` function in R then displays a table of values for the first six genes.

```{r}
# Read in data (tab separated)
gene_data <- read.table("gene_expression.tsv", header = TRUE, row.names = 1, sep = "\t")

# Show first six rows
head(gene_data)
```

#### 1.2. Add mean column

We make a new column which is the mean of the other columns. Again, the first six genes is displayed through the `head()` function.

```{r}
# Create new column with mean of other columns
gene_data$mean <- rowMeans(gene_data)

# Display first 6
head(gene_data)
```

#### 1.3. List top 10 genes with highest mean expression

This selects the top 10 genes with the highest mean expression values from the `gene_data` dataframe and stores them in `top_genes`. The '-' sign before `gene_data$mean` tells it to sort in decreasing order.

```{r}
# Create variable to store top 10 genes with highest mean expression and sort in descending order
top_genes <- head(gene_data[order(-gene_data$mean), ], 10)

# Print out top 10 genes
top_genes
```

#### 1.4. Determine the number of genes with a mean <10

Create variable `low_mean_count` and store the total number of mean values <10. When returning the `low_mean_count` value, we receive *35,988 genes* with a mean <10.

```{r}
# Create variable for data with mean values less than 10
low_mean_count <- sum(gene_data$mean < 10)

# Print out low mean count
low_mean_count
```

#### 1.5. Make a histogram plot of the mean values and include it into your report.

The below creates a histogram to visualise distribution of gene expression levels across all genes. A peak near lower values indicates many genes have low expression.

```{r}
# Create histogram using the mean column
hist(gene_data$mean, 
     breaks = 100, # Divide the x-axis into 100 bins
     xlim = c(0, 550000), # Set the range of the x-axis (mean expression values)
     ylim = c(0, 60000), # Set the range of the y-axis (frequency of genes)
     main = "Histogram of Mean Gene Expression", # Title of the histogram
     xlab = "Mean Expression") # Label for the x-axis
```


### Growth Data Analysis

#### 1.6. Import this csv file into an R object. What are the column names?

The column names are printed using `colnames()` as follows. They are: "Site", "TreeID", "Circumf_2005_cm", "Circumf_2010_cm", "Circumf_2015_cm", and "Circumf_2020_cm".

```{r}
#store csv into a variable
growth_data <- read.csv("growth_data.csv", header = TRUE)

#print out column names
colnames(growth_data)
```

#### 1.7. Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.

The sites available are either "northeast" or "southwest". Our columns suggest the start of the study period is 2005 and the end of the study is 2020. The below data frame contains a calculation of the mean and standard deviation for both sites at 2005 and 2020.

```{r}
#create a data frame for the statistics of mean and sd
stats <- data.frame(
  Site = unique(growth_data$Site),
  mean_2005 = tapply(growth_data$Circumf_2005_cm, growth_data$Site, mean),
  sd_2005   = tapply(growth_data$Circumf_2005_cm, growth_data$Site, sd),
  mean_2020 = tapply(growth_data$Circumf_2020_cm, growth_data$Site, mean),
  sd_2020   = tapply(growth_data$Circumf_2020_cm, growth_data$Site, sd)
)

#print out data frame for statistics containing the mean and sd
stats
```

#### 1.8. Make a box plot of tree circumference at the start and end of the study at both sites.

The boxplot shows that the tree circumference at the end for both sites grew, with the northeast site showing more growth than the southwest site at the end of the study in 2020.

```{r}
# Create combined vector of circumference values
circumference <- c(growth_data$Circumf_2005_cm, growth_data$Circumf_2020_cm)

# Create corresponding Site labels
site <- rep(growth_data$Site, 2)

# Create corresponding Year labels
year <- c(rep("2005", nrow(growth_data)), rep("2020", nrow(growth_data)))

# Combine Site and Year into one factor
site_year <- interaction(site, year)

# Make nicer labels
levels(site_year) <- c("Northeast 2005", "Southwest 2005",
                       "Northeast 2020", "Southwest 2020")

# Box plot
boxplot(circumference ~ site_year,
        main = "Tree Circumference at Start (2005) and End (2020)",
        xlab = "Site and Year",
        ylab = "Circumference (cm)",
        col = c("goldenrod", "cornflowerblue", "goldenrod", "cornflowerblue"))
```

#### 1.9. Calculate the mean growth over the last 10 years at each site.

The mean growth over the last 10 years at each site is:

* Northeast site: **42.94 cm**
* Southwest site: **35.49 cm**

```{r}
# Calculate 10-year growth per tree using 2020 and 2010
growth_data$growth10yr <- growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm

# Mean growth per site
growth_summary <- aggregate(growth10yr ~ Site, data = growth_data, mean)

#Print out growth summary
growth_summary 
```

We use `diff` on the `growth10yr` column to subtract the values. So, on average, trees at the northeast site grew **~7.45cm** more over the last 10 years than trees at the southwest site.

```{r}
# subtract the growth10yr values from each other
diff(growth_summary$growth10yr)
```

#### 1.10. Use the t.test to estimate the p-value that the 10 year growth is different at the two sites.

A Welch two-sample t-test indicated that this difference was not statistically significant (t = 1.8882, df = 87.978, p = 0.06229), with a 95% confidence interval of -0.39 to 15.29 cm. The p-value is slightly above the common significance threshold of 0.05.

```{r}
t.test(growth10yr ~ Site, data = growth_data)
```

## 2: Examining biological sequence diversity

**Organisms for comparison**

In this section we are studying **Acetobacter aceti (GCA_002005445)** [@ensembl_aaceti] and comparing its sequence features to **E. coli (GCA_000005845)** [@ensembl_ecoli].

* **Acetobacter aceti:** Gram-negative bacterium known for production of acetic acid and used for converting ethanol into vinegar (acetic acid) [@Gomes2018].
* **Escherichia coli:** Gram-negative bacterium. E. coli can ferment glucose to produce acetate, but this growth is inhibited by high acetate concentrations. It plays a role in the gut microbiota and in food production, but is also a common pathogen, causing a range of diseases [@Jang2017].

#### 2.1. Genome and CDS analysis

*Download the whole set of coding DNA sequences for E. coli and your organism of interest. How many coding sequences are present in these organisms? Present this in the form of a table. Describe any differences between the two organisms.*

##### 2.1.1. Download the whole set of coding DNA sequences for E. coli and your organism of interest.

Download both E. coli (GCA_000005845) and A. aceti (GCA_002005445) from Ensembl Bacteria.

```{r}
# E. coli file download
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")

# unzip and allow for overwrite if existing file
gunzip("ecoli_cds.fa.gz", overwrite = TRUE)

# A. aceti file download
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_80_collection/acetobacter_aceti_gca_002005445/cds/Acetobacter_aceti_gca_002005445.ASM200544v1.cds.all.fa.gz"
download.file(URL,destfile="aaceti_cds.fa.gz")

# unzip and allow for overwrite if existing file
gunzip("aaceti_cds.fa.gz", overwrite = TRUE)

# display files in current folder to confirm download
list.files()
```

##### 2.1.2. How many coding sequences are present in these organisms? Present this in the form of a table.

We can figure out the number of coding sequences (CDS) and present this in a table called `cds_table`.

```{r}
# Load coding sequences (CDS) from FASTA files using the seqinr package
ecoli_cds <- seqinr::read.fasta("ecoli_cds.fa")
aaceti_cds <- seqinr::read.fasta("aaceti_cds.fa")

# Create summary data frame
cds_table <- data.frame(
  Organism = c("E. coli", "A. aceti"), # Names of the organisms
  CDS_Count = c(length(ecoli_cds), length(aaceti_cds)) # Number of CDS entries in each organism
)

# Print CDS count table
cds_table

# Calculate the difference in CDS count between the two organisms
# diff() returns the difference
diff(cds_table$CDS_Count)
```
##### 2.1.3. Describe any differences between the two organisms.

As per the table, E. coli has **837** more coding sequences than A. aceti (4239 - 3402 = 837).

#### 2.2. Total coding DNA

*How much coding DNA is there in total for these two organisms? Present this in the form of a table. Describe any differences between the two organisms.*

##### 2.2.1 How much coding DNA is there in total for these two organisms? Present this in the form of a table.

We want all rows and first column in the cds files to be summed. The table shows total Coding DNA in E. coli and A. aceti.

```{r}
# Total coding DNA from the length column
ecoli_len <- as.numeric(summary(ecoli_cds)[,1])
aaceti_len <- as.numeric(summary(aaceti_cds)[,1])

# Sum of all the lengths within the organism 
ecoli_total <- sum(ecoli_len)
aaceti_total <- sum(aaceti_len)

# Create table
DNA_total <- data.frame(
  Organism = c("E. coli", "A. aceti"),
  Total_DNA = c(ecoli_total, aaceti_total)
)

# Print total coding DNA table
DNA_total

# Difference in Total_DNA to show which organism has more coding DNA in absolute terms
diff(DNA_total$Total_DNA)
```

##### 2.2.2. Describe any differences between the two organisms.

The total coding DNA in E. coli is about **741,960** base pairs greater than in A. aceti. This difference is consistent with its larger number of coding genes as noted in question 1. This could suggest E. coli likely has a larger and more diverse genome than A. aceti.

#### 2.3. Coding sequence length, visualisation and analysis

*Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms. What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms.*

##### 2.3.1. Calculate the length of all coding sequences in these two organisms. What is the mean and median coding sequence length of these two organisms? 

*Length of all coding sequences* as per Question 2 above (total DNA).

Calculate the mean and median coding sequences then store it all in a data frame.

```{r}
# Calculate mean and median CDS length
cds_stats <- data.frame(
  Organism = c("E. coli", "A. aceti"),
  Length = c(ecoli_total, aaceti_total),
  Mean_CDS_Length = c(mean(ecoli_len), mean(aaceti_len)),
  Median_CDS_Length = c(median(ecoli_len), median(aaceti_len))
)

# Print table to show count, means and medians
cds_stats

#Calculate differences in mean and median coding sequences
diff(cds_stats$Mean_CDS_Length)
diff(cds_stats$Median_CDS_Length)
```

##### 2.3.2. Make a boxplot of coding sequence length in these organisms.

This is a boxplot for the CDS lengths of both organisms.

```{r}
# Boxplot of CDS lengths
boxplot(list("E. coli" = ecoli_len, "A. aceti" = aaceti_len),
        main = "Distribution of coding sequence lengths (box plot)",
        ylab = "Sequence Length (bp)",
        col = c("skyblue", "plum1"))
```

##### 2.3.3. Describe any differences between the two organisms.

* A. aceti coding sequences are longer on average than those in E. coli (approximately **12 bp** difference) but this is very small relative to the overall sequence length and might not be biologically significant.

* The median coding sequence length is longer in E. coli (**831 bp vs 819 bp**), which suggests that E. coli has a slightly more balanced distribution of gene lengths. A. aceti may have a few longer genes that increase its mean (outliers can be seen in the boxplot).

* The fact that **A. aceti has a higher mean but a lower median** implies its coding sequence length distribution might be more skewed, with some unusually long genes pulling the mean upward as visualised in our boxplot.


#### 2.4. Nucleotide and amino acid frequency

*Calculate the frequency of DNA bases in the total coding sequences for both organisms. Perform the same calculation for the total protein sequence. Create bar plots for nucleotide and amino acid frequency. Describe any differences between the two organisms.*

##### 2.4.1. Calculate the frequency of DNA bases in the total coding sequences for both organisms.

Use `unlist` to count the entire genome.

```{r}
# Concatenate all CDS into one sequence per organism
ecoli_seq <- unlist(ecoli_cds)
aaceti_seq <- unlist(aaceti_cds)

# Count combinations of 1s of the nucleotides in entire DNA using seqinr::count
ecoli_dna_composition <- seqinr::count(ecoli_seq,1)
aaceti_dna_composition <- seqinr::count(aaceti_seq,1)

#Print nucleotide composition
ecoli_dna_composition
aaceti_dna_composition

# Relative frequencies can also be calculated using freq = TRUE
ecoli_nt_freq <- seqinr::count(ecoli_seq,1, freq = TRUE)
aaceti_nt_freq <- seqinr::count(aaceti_seq,1, freq = TRUE)

# Print relative frequencies for E. coli
ecoli_nt_freq 
# Print relative frequencies for A. aceti
aaceti_nt_freq
```
##### 2.4.2. Create bar plots for nucleotide frequency

Using the frequencies stored in `ecoli_nt_freq` and `aaceti_nt_freq`, group barplots to visualise the nucleotide composition.

```{r}
# Combine data into a matrix
nt_matrix <- rbind(ecoli_nt_freq, aaceti_nt_freq)
colnames(nt_matrix) <- names(ecoli_nt_freq)
rownames(nt_matrix) <- c("E. coli", "A. aceti")

# Grouped barplot
barplot(nt_matrix, beside = TRUE, 
        col = c("skyblue", "plum1"),
        xlab = "Nucleotides",
        ylab = "Proportion",
        main = "CDS Nucleotide Composition",
        ylim = c(0, max(nt_matrix)*1.2))  # add 20% extra space on top
legend("topright", legend = rownames(nt_matrix), fill = c("skyblue", "plum1"))
grid()
```

##### 2.4.3. Perform the same calculation for the total protein sequence

In the code, `lapply` applies the translate function to each element of the cds list, resulting in prot, a list of translated protein sequences.

```{r}
#unlist and translate entire genome
ecoli_prot <- unlist(lapply(ecoli_cds, translate))
aaceti_prot <- unlist(lapply(aaceti_cds, translate))
```

Here we store E. coli unique amino acids in `aa` and A. aceti in `aa1`. Both have 20 unique amino acids.

```{r}
# Find unique number of amino acids in E. coli
aa <- unique(ecoli_prot)
aa <- aa[aa != "*"] # removes any asterisks for unknown amino acids
length(aa) # counts the remaining unique amino acids

# Find unique number of amino acids in A. aceti
aa1 <- unique(aaceti_prot)
aa1 <- aa1[aa1 != "*"]
length(aa1) # counts the remaining unique amino acids
```

We use `aa` and `aa1` to determine the count of each amino acid coded by E. coli and A. aceti, respectively. The following counts the occurence of each amino acid using the total set of possible amino acids.

```{r}
# Count amino acids using seqinr::count
ecoli_aa_count <- seqinr::count(ecoli_prot,wordsize=1,alphabet=aa)
aaceti_aa_count <- seqinr::count(aaceti_prot,wordsize=1,alphabet=aa1)

# Print total amino acid count
ecoli_aa_count
aaceti_aa_count
```

##### 2.4.4. Create bar plots for amino acid frequency

We convert the count to relative frequencies in order to compare their composition with the same y-axis.

##### Barplot side-by-side comparison of amino acid proportions

We do a side-by-side comparison of the two amino acid compositions using their relative frequencies. While both organisms display broadly similar distributions, we can see A. aceti shows higher proportions of alanine (A), glycine (G), proline (P), arginine (R), and serine (S), whereas E. coli is enriched in isoleucine (I), lysine (K), asparagine (N), and glutamine (Q).

```{r}
# Convert to proportions
ecoli_aa_proportion <- seqinr::count(ecoli_prot,wordsize=1,alphabet=aa, freq = TRUE)
aaceti_aa_proportion <- seqinr::count(aaceti_prot,wordsize=1,alphabet=aa1, freq = TRUE)

# Combine proportions into a matrix
aa_matrix <- rbind(E_coli = as.numeric(ecoli_aa_proportion),
                   A_aceti = as.numeric(aaceti_aa_proportion))
colnames(aa_matrix) <- names(ecoli_aa_proportion)  # amino acid labels

# Side-by-side barplot
barplot(aa_matrix, 
        beside = TRUE, 
        col = c("skyblue", "plum1"),
        xlab = "Amino Acid",
        ylab = "Proportion",
        main = "Amino Acid Proportion Comparison",
        ylim = c(0, max(aa_matrix)*1.2))  # extra space for legend

# Add legend
legend("topright", legend = rownames(aa_matrix), fill = c("skyblue", "plum1"))

# Add gridlines
grid()
```

##### Table to review nucleotide and amino acid proportions

To compare the two organisms' nucleotides and amino acid sequences, we use `knitr` to generate a clean html table. The table contains a column for both the organisms' count and proportions.

```{r}
# DNA nucleotide frequency table with counts
nt_freq_table <- data.frame(
  Base = union(names(ecoli_nt_freq), names(aaceti_nt_freq)),
  E_coli_Count = as.numeric(ecoli_dna_composition[union(names(ecoli_nt_freq), names(aaceti_nt_freq))]),
  A_aceti_Count = as.numeric(aaceti_dna_composition[union(names(ecoli_nt_freq), names(aaceti_nt_freq))]),
  E_coli_Proportion = as.numeric(ecoli_nt_freq[union(names(ecoli_nt_freq), names(aaceti_nt_freq))]),
  A_aceti_Proportion = as.numeric(aaceti_nt_freq[union(names(ecoli_nt_freq), names(aaceti_nt_freq))])
)

# Replace NAs with 0
nt_freq_table[is.na(nt_freq_table)] <- 0

# Render table
knitr::kable(nt_freq_table, digits = 4,
             col.names = c("Base",
                           "E. coli (Count)", "A. aceti (Count)",
                           "E. coli (Proportion)", "A. aceti (Proportion)"),
             caption = "Nucleotide Counts and Proportions for E. coli and A. aceti")

# Amino acid frequency table with counts
aa_freq_table <- data.frame(
  AminoAcid = union(names(ecoli_aa_proportion), names(aaceti_aa_proportion)),
  E_coli_Count = as.numeric(ecoli_aa_count[union(names(ecoli_aa_proportion), names(aaceti_aa_proportion))]),
  A_aceti_Count = as.numeric(aaceti_aa_count[union(names(ecoli_aa_proportion), names(aaceti_aa_proportion))]),
  E_coli_Proportion = as.numeric(ecoli_aa_proportion[union(names(ecoli_aa_proportion), names(aaceti_aa_proportion))]),
  A_aceti_Proportion = as.numeric(aaceti_aa_proportion[union(names(ecoli_aa_proportion), names(aaceti_aa_proportion))])
)

# Replace NAs with 0
aa_freq_table[is.na(aa_freq_table)] <- 0

# Generate table with title
knitr::kable(aa_freq_table, digits = 4,
             col.names = c("Amino Acid", 
                           "E. coli (Count)",
                           "A. aceti (Count)",
                           "E. coli (Proportion)", 
                           "A. aceti (Proportion)"),
             caption = "Amino Acid Counts and Proportions for E. coli and A. aceti")
```

##### 2.4.5. Describe any differences between the two organisms.

**Nucleotides**

From our results regarding the frequency of DNA bases, we can see that E. coli and A. aceti differ noticeably in base composition.

* Comparison of nucleotide composition shows that E. coli has a relatively balanced AT/GC content (around 48% AT, 52% GC)
* A. aceti is more GC-rich (around 42% AT, 58% GC).
* This indicates that A. aceti has a higher GC content compared to E. coli, which could potentially influence genome stability, codon usage (explored later in question 5) and possibly its gene expression [@Agashe2014].

**Amino Acid**

* Both E. coli and A. aceti show broadly similar amino acid usage.
* Both organisms have Leucine (L) as the most abundant amino acid (~10%) followed by Alanine (A) and Glycine (G).
* Cysteine (C) and Tryptophan (W) are rare in both (<2%).
* Differences are A. aceti shows higher proportions of alanine (A: 0.1098 vs 0.0954), glycine (G: 0.0831 vs 0.0736), proline (P: 0.0525 vs 0.0444), arginine (R: 0.0692 vs 0.0553), and serine (S: 0.0642 vs 0.0578).
* In contrast, E. coli is enriched in isoleucine (I: 0.0601 vs 0.0524), lysine (K: 0.0440 vs 0.0325), asparagine (N: 0.0390 vs 0.0287), and glutamine (Q: 0.0445 vs 0.0332). These differences likely reflect adaptations to the metabolic and/or environmental requirements of each organism.

#### 2.5. Codon usage bias

*Create a codon usage table and quantify the codon usage bias among all coding sequences. Describe any differences between the two organisms with respect to their codon usage bias. Provide charts to support your observations.*

##### 2.5.1. Create a codon usage table and quantify the codon usage bias among all coding sequences
A codon is a three base triplet that codes for a particular amino acid. As there are 64 possible codons and just 20 amino acids (and 3 termination codons); some amino acids are coded by multiple codons. For example, leucine is coded by six codons that are synonymous for it. `seqinr` has a command called `uco` which calculates codon usage information.

Relative Synonymous Codon Usage (RSCU) (called via `index=rscu`) is a measure as to how preferred a codon of a particular amino acid is compared to the other codons which also encode it.

* RSCU = 1 means codon used as expected.
* RSCU > 1 means codon is preferred.
* RSCU < 1 means codon is under-represented.

We create a codon usage table to quantify the codon usage bias for both E. coli and A. aceti.
```{r}
# Codon usage tables
# Concatenate all CDS into one long sequence and get codon frequencies
ecoli_codon <- uco(unlist(ecoli_cds),index="rscu", as.data.frame = TRUE)
aaceti_codon <- uco(unlist(aaceti_cds), index = "rscu", as.data.frame = TRUE)
```

Combine `ecoli_codon` and `aaceti_codon` into one table with the amino acid and codon columns called `rscu_table`.

```{r}
# Make sure RSCU columns have unique names
colnames(ecoli_codon)[colnames(ecoli_codon) == "RSCU"] <- "RSCU_Ecoli"
colnames(aaceti_codon)[colnames(aaceti_codon) == "RSCU"] <- "RSCU_Aaceti"
# Make sure frequency columns have unique names
colnames(ecoli_codon)[colnames(ecoli_codon) == "freq"] <- "freq_Ecoli"
colnames(aaceti_codon)[colnames(aaceti_codon) == "freq"] <- "freq_Aaceti"

# Merge by Codon (can use either AA column; as they should match)
rscu_table <- merge(ecoli_codon[, c("codon", "AA", "RSCU_Ecoli", "freq_Ecoli")],
                    aaceti_codon[, c("codon", "RSCU_Aaceti", "freq_Aaceti")],
                    by = "codon")

# Reorder columns
rscu_table <- rscu_table[, c("codon", "AA", "freq_Ecoli","RSCU_Ecoli", "freq_Aaceti", "RSCU_Aaceti")]

# Print entire RSCU table with knitr::kable
knitr::kable(rscu_table, 
             caption = "RSCU and codon frequencies for E. coli and A. aceti", # Table title
             digits = 4) # rounds values to these decimal places
```


##### 2.5.2. Provide charts to support your observations.

The following charts are provided for both organisms:

* Codon Usage Bias Heatmap
* Stacked Bar and Tile Charts for Codon Usage
* Bar Plots per Amino Acid and RSCU Values

**Codon Usage Bias Heatmap (RSCU values)**

To further assist with visualisation of codon usage bias in both organisms, the below creates a heatmap for codon bias using `ggplot2`. Here the codons are aligned on the x-axis, with the organisms in the y-axis. We can compare the RSCU visually this way.

```{r}
# add column called Organism to allow for labels
ecoli_codon$Organism <- "E. coli"
aaceti_codon$Organism <- "A. aceti"

# Create simplified data frames for heatmap with relevant columns
df_ecoli <- data.frame(
  Codon = ecoli_codon$codon, # codon column from ecoli_codon
  RSCU  = ecoli_codon$RSCU_Ecoli, # RSCU values for E. coli
  Organism = "E. coli"  # add label
)

df_aaceti <- data.frame(
  Codon = aaceti_codon$codon, # codon column from aaceti_codon
  RSCU  = aaceti_codon$RSCU_Aaceti, # RSCU values for A. aceti
  Organism = "A. aceti" # add label
)

# Combine both data frames
rscu_all <- rbind(df_ecoli, df_aaceti)

# Plot heat map with ggplot2
ggplot(rscu_all, aes(x = Codon, y = Organism, fill = RSCU)) +
  geom_tile(color = "white") +
  scale_fill_gradient2(low = "blue4", mid = "white", high = "deeppink4",
                       midpoint = 1, name = "RSCU") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "Codon Usage Bias Heatmap (RSCU values)",
       x = "Codon", y = "Organism")

```

**Codon Usage Bias: Stacked Bar and Tile Charts**

To better understand codon usage patterns, we grouped codons by their corresponding amino acids and calculated their relative synonymous codon usage (RSCU). The stacked bar charts highlight how frequently each codon is used for a given amino acid, making codon bias across amino acids clearer.

In addition, tile plots below each bar chart show all codons that encode the same amino acid. This provides a direct mapping between codons and their relative contributions within each amino acid group.

Note: The charts display 21 amino acids, where **Stp** represents the stop codons (TAA, TAG and TGA).

```{r}
# Extract Codon, RSCU values and Amino Acid (AA) from rscu_table
dat_aaceti <- data.frame(
  Codon = rscu_table$codon,
  RSCU  = rscu_table$RSCU_Aaceti,
  AA    = rscu_table$AA
)

dat_ecoli <- data.frame(
  Codon = rscu_table$codon,
  RSCU  = rscu_table$RSCU_Ecoli,
  AA    = rscu_table$AA
)

# For each AA, assign codons sequential group IDs.
# This ensures all codons for the same amino acid get related colour categories (used consistently across plots)
dat_aaceti$Col <- 1
dat_ecoli$Col <- 1

# This populates the Col column with unique values for each codon coding an amino. Each value is a colour, therefore all amino acids will share the same values but the codons will be different.
dat_aaceti <- dat_aaceti %>%
  group_by(AA) %>%
  mutate(Col = lag(cumsum(Col), default = 0),
         Col = as.factor(Col))

dat_ecoli <- dat_ecoli %>%
  group_by(AA) %>%
  mutate(Col = lag(cumsum(Col), default = 0),
         Col = as.factor(Col))

# Create bar plots - A. aceti
p0 <- ggplot(data=dat_aaceti, aes(x=AA, y=RSCU, fill=Col)) +
  geom_bar(stat='identity') +
  theme_minimal() +
  theme(axis.title.x = element_blank()) +
  guides(fill = "none") +
  ggtitle("Codon Usage Bias in A. aceti") +
  theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))

# Create tile plot - A. aceti
p1 <- ggplot(data=dat_aaceti) +
  geom_tile(aes(x=AA, y=fct_rev(Col), fill=Col), col="white") +
  geom_text(aes(x=AA, y=Col, label=Codon), col="white", size=3.5) +
  theme_void() +
  theme(legend.position="none") +
  guides(fill = "none")

# Create bar plots - E. coli
p2 <- ggplot(data=dat_ecoli, aes(x=AA, y=RSCU, fill=Col)) +
  geom_bar(stat='identity') +
  theme_minimal() +
  theme(axis.title.x = element_blank()) +
  guides(fill = "none") +
  ggtitle("Codon Usage Bias in E. coli") +
  theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))

# Create tile plot - E. coli
p3 = ggplot(data=dat_ecoli) +
  geom_tile(aes(x=AA, y=fct_rev(Col), fill=Col), col="white") +
  geom_text(aes(x=AA, y=Col, label=Codon), col="white", size=3.5) +
  theme_void() +
  theme(legend.position="none") +
  guides(fill = "none")

# Bar plot (7 units tall) stacked on tile plot (3 unit tall)
plot_aaceti <- (p0 / p1) + plot_layout(heights = c(7, 3))
plot_ecoli  <- (p2 / p3) + plot_layout(heights = c(7, 3))

# Print plots 
plot_aaceti   # RSCU codon usage stacked bar plot for A. aceti
plot_ecoli    # RSCU codon usage stacked bar plot for E. coli
```

**Codon Usage Bias: Bar Plots per Amino Acid (RSCU Values) - A. aceti**

We can also create individual bar plots for each amino acid. Each bar shows the RSCU for the codons encoding that amino acid, allowing us to see which codons are preferred.

These plots complement the earlier stacked and tile charts by breaking down each amino acid into separate visualisations, making it easier to compare codon bias across amino acids.

For clarity, stop codons (**Stp**) are excluded from these plots so that only the 20 standard amino acids are shown.

```{r}
## Store data into a new data frame structure
aaceti_dat <- data.frame(
  Codon = rscu_table$codon,
  RSCU  = rscu_table$RSCU_Aaceti,
  AA    = rscu_table$AA,
  stringsAsFactors = FALSE
)

# Remove stop codons
aaceti_dat <- aaceti_dat[aaceti_dat$AA != "Stp", ]

# Order by AA then Codon
aaceti_dat <- aaceti_dat[order(aaceti_dat$AA, aaceti_dat$Codon), ]
aa_levels <- unique(aaceti_dat$AA)

# Set plotting area: adjust nrow/ncol depending on number of amino acids
nAA <- length(aa_levels)
ncol <- 5   # choose grid layout
nrow <- ceiling(nAA / ncol)

par(mfrow=c(nrow, ncol))
par(mar=c(3,3,2,1), oma=c(3,1,3,1), cex.axis=0.8, cex.lab=0.8, cex.main=0.9)

# Loop through each AA and plot codon usage
for (current_aa in aa_levels) {
  subset_aaceti_dat <- aaceti_dat[aaceti_dat$AA == current_aa, ]
  
  bp <- barplot(
    subset_aaceti_dat$RSCU,
    names.arg = subset_aaceti_dat$Codon,
    ylim = c(0, max(aaceti_dat$RSCU) + 0.5),
    col = "plum1",
    las = 2,
    cex.names = 0.8,
    ylab = "RSCU",
    main = current_aa
  )
  
  # Reference line at RSCU = 1
  abline(h=1, col="red", lty=2)
}

# Add a main title for the entire figure
mtext("Codon Usage Bias Across Amino Acids (A. aceti)", outer = TRUE, cex = 0.9, font = 2)
```

**Codon Usage Bias: Bar Plots per Amino Acid (RSCU Values) - E. coli**

```{r}
## Store data into a new data frame structure
ecoli_dat <- data.frame(
  Codon = rscu_table$codon,
  RSCU  = rscu_table$RSCU_Ecoli,
  AA    = rscu_table$AA,
  stringsAsFactors = FALSE
)

# Remove stop codons
ecoli_dat <- ecoli_dat[ecoli_dat$AA != "Stp", ]

# Order by AA then Codon
ecoli_dat <- ecoli_dat[order(ecoli_dat$AA, ecoli_dat$Codon), ]
aa_levels_1 <- unique(ecoli_dat$AA)

# Set plotting area: adjust nrow/ncol depending on number of amino acids
nAA <- length(aa_levels_1)
ncol <- 5   # choose grid layout
nrow <- ceiling(nAA / ncol)

par(mfrow=c(nrow, ncol))
par(mar=c(3,3,2,1), oma=c(3,1,3,1), cex.axis=0.8, cex.lab=0.8, cex.main=0.9)

# Loop through each AA and plot codon usage
for (current_aa_1 in aa_levels_1) {
  subset_ecoli_dat <- ecoli_dat[ecoli_dat$AA == current_aa_1, ]
  
  bp1 <- barplot(
    subset_ecoli_dat$RSCU,
    names.arg = subset_ecoli_dat$Codon,
    ylim = c(0, max(ecoli_dat$RSCU) + 0.5),
    col = "skyblue",
    las = 2,
    cex.names = 0.8,
    ylab = "RSCU",
    main = current_aa_1
  )
  
  # Reference line at RSCU = 1
  abline(h=1, col="red", lty=2)
}

# Add a main title for the entire figure
mtext("Codon Usage Bias Across Amino Acids (E. coli)", outer = TRUE, cex = 0.9, font = 2)
```

##### 2.5.3. Describe any differences between the two organisms with respect to their codon usage bias. 

Both organisms exhibit codon usage bias, with some codons for the same amino acid used more frequently in one organism than the other

* The CTG codon (codes for Leu) is most overrepresented in both organisms as shown in the heatmap, the codon usage stacked bargraph and amino acid barplots.
* Analysis of RSCU values shows that E. coli exhibits strong codon usage bias, favoring specific codons for several amino acids (e.g., AAA for Lys, CCG for Pro, CTG for Leu) while underusing others.
* A. aceti displays a more balanced codon usage, for example, AAG is slightly preferred over AAA for Lys, CCC over CCG for Pro, and TGA is the dominant stop codon instead of TAA.
* Other amino acids, such as Gly and Ala, show similar patterns in both organisms.
* E. coli tends to have more extreme biases as clearly shown on our heatmap.
* The differences may reflect organism-specific adaptations in protein structure, metabolic needs and protein translational efficiency. Our side-by-side barplots of amino acid proportions highlight these subtle but consistent shifts between the two species [@Bulmer1991].

#### 2.6. K-mer analysis 

In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most over- and under-represented k-mers in your organism of interest. Are these k-mers also over- and under-represented in E. coli to a similar extent? Provide plots to support your observations. Why do you think these sequences are present at different levels in the genomes of these organisms?

##### 2.6.1. Finding the top 10 over and under represented k-mers in A. aceti for length 3-5

Calculate the frequency of amino acid k-mers (short subsequences of length k) in the translated proteomes of E. coli and A. aceti. By counting k-mers of length 3, 4, and 5, and then converting counts to relative frequencies, we can compare patterns of over- and under-representation between organisms while controlling for differences in overall proteome size.

```{r}
# unlist protein sequence
ecoli_kmer <- unlist(ecoli_prot)
aaceti_kmer <- unlist(aaceti_prot)

# Calculate k-mer frequencies for each organism
# Using freq = TRUE ensures we get relative frequencies 
# 3-mer frequencies
ecoli_k3_freq <- seqinr::count(ecoli_kmer,wordsize=3,alphabet=aa,freq=TRUE)
aaceti_k3_freq <- seqinr::count(aaceti_kmer,wordsize=3,alphabet=aa1,freq=TRUE)

# 4-mer frequencies
ecoli_k4_freq <- seqinr::count(ecoli_kmer,wordsize=4,alphabet=aa,freq=TRUE)
aaceti_k4_freq <- seqinr::count(aaceti_kmer,wordsize=4,alphabet=aa1,freq=TRUE)

# 5-mer frequencies
ecoli_k5_freq <- seqinr::count(ecoli_kmer,wordsize=5,alphabet=aa,freq=TRUE)
aaceti_k5_freq <- seqinr::count(aaceti_kmer,wordsize=5,alphabet=aa1,freq=TRUE)
```

Create frequency data frames for each k-mer size. Each table stores the relative frequencies of all possible k-mers for both organisms side by side (E. coli vs. A. aceti). This makes it easy to compare the distribution of k-mers across proteomes.

```{r}
# 3-mer frequency table
k3_table <- data.frame(
    Ecoli = ecoli_k3_freq,
  Aaceti = aaceti_k3_freq
)

# 4-mer frequency table
k4_table <- data.frame(
    Ecoli = ecoli_k4_freq,
  Aaceti = aaceti_k4_freq
)

# 5-mer frequency table
k5_table <- data.frame(
    Ecoli = ecoli_k5_freq,
  Aaceti = aaceti_k5_freq
)
```

We create a ranking column to label the top/bottom A. aceti and E. coli kmers. The ranking of both organisms mean we can create rank-rank scatterplots.

```{r}
#Rank the organisms for k=3 and store in a new column
k3_table$Aaceti.rank <- rank(-k3_table$Aaceti.Freq, ties.method = "min")
k3_table$Ecoli.rank <- rank(-k3_table$Ecoli.Freq, ties.method = "min")

#Rank the organisms for k=4 and store in a new column
k4_table$Aaceti.rank <- rank(-k4_table$Aaceti.Freq, ties.method = "min")
k4_table$Ecoli.rank <- rank(-k4_table$Ecoli.Freq, ties.method = "min")

#Rank the organisms for k=5 and store in a new column
k5_table$Aaceti.rank <- rank(-k5_table$Aaceti.Freq, ties.method = "min")
k5_table$Ecoli.rank <- rank(-k5_table$Ecoli.Freq, ties.method = "min")
```

We can then get the top 10 highest and lowest frequency kmers and confirm this with the ranking column. We can also view where E. coli's corresponding kmers rank.

```{r}
#for k=3
# Sort descending to get top 10 highest
top10_k3 <- head(k3_table[order(-k3_table$Aaceti.Freq), ], 10)
# Sort ascending to get top 10 lowest
bottom10_k3 <- head(k3_table[order(k3_table$Aaceti.Freq), ], 10)
# Combine into a single table with an extra column to indicate high/low
k3_summary <- rbind(
  data.frame(Type = "Highest", top10_k3),
  data.frame(Type = "Lowest", bottom10_k3)
)

# for k=4
# Sort descending to get top 10 highest
top10_k4 <- head(k4_table[order(-k4_table$Aaceti.Freq), ], 10)
# Sort ascending to get top 10 lowest
bottom10_k4 <- head(k4_table[order(k4_table$Aaceti.Freq), ], 10)
# Combine into a single table with an extra column to indicate high/low
k4_summary <- rbind(
  data.frame(Type = "Highest", top10_k4),
  data.frame(Type = "Lowest", bottom10_k4)
)

# for k=5
# Sort descending to get top 10 highest
top10_k5 <- head(k5_table[order(-k5_table$Aaceti.Freq), ], 10)
# Sort ascending to get top 10 lowest
bottom10_k5 <- head(k5_table[order(k5_table$Aaceti.Freq), ], 10)
# Combine into a single table with an extra column to indicate high/low
k5_summary <- rbind(
  data.frame(Type = "Highest", top10_k5),
  data.frame(Type = "Lowest", bottom10_k5)
)

# Create a table using knitr::kable for better html output
# table for k=3
knitr::kable(
  k3_summary[, c("Type", "Aaceti.Var1", "Aaceti.Freq", "Ecoli.Freq", "Aaceti.rank", "Ecoli.rank")], 
  caption = "Top 10 highest and lowest frequency 3-mers in A. aceti",
  digits = 6 # decimal places
)

# table for k=4
knitr::kable(
  k4_summary[, c("Type", "Aaceti.Var1", "Aaceti.Freq", "Ecoli.Freq", "Aaceti.rank", "Ecoli.rank")], 
  caption = "Top 10 highest and lowest frequency 4-mers in A. aceti",
  digits = 6 # decimal places
)

# table for k=5
knitr::kable(
  k5_summary[, c("Type", "Aaceti.Var1", "Aaceti.Freq", "Ecoli.Freq", "Aaceti.rank", "Ecoli.rank")], 
  caption = "Top 10 highest and lowest frequency 5-mers in A. aceti",
  digits = 6 # decimal places
)
```

##### 2.6.2. Provide plots to support your observations.

The following plots are provided:

* Rank/rank scatterplots
* Barplots of 10 most frequent and 10 least frequent k-mers in A. aceti

###### Rank/rank scatterplots

By ranking k-mer frequencies above, we can directly compare how similarly A. aceti and E. coli use particular motifs. A useful way to visualise this is with a **rank–rank scatterplot**, which plots the rank of each k-mer in one organism against its rank in the other. If a k-mer has a similar ranking in both organisms, the point will fall close to the red diagonal line. For instance, if the top-ranked k-mer in E. coli is also the top-ranked in A. aceti, it will lie exactly on the line. Points that deviate strongly from the line highlight k-mers with very different relative usage between the two proteomes. We will generate such scatterplots for k-mers of length 3, 4, and 5.

```{r}
ggplot(k3_table, aes(x = Ecoli.rank, y = Aaceti.rank)) +
  geom_point(alpha = 0.6, color = "royalblue3") +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "red") +
  labs(title = "Rank correlation of 3-mers",
       x = "E. coli Rank",
       y = "A. aceti Rank") +
  theme(plot.title = element_text(hjust = 0.5))
```

The rank/rank scatterplot for 4-mer shows that there is moderate correlation between ranks. Rank correlation is less strong than that of the shorter 3-mer.

```{r}
ggplot(k4_table, aes(x = Ecoli.rank, y = Aaceti.rank)) +
  geom_point(alpha = 0.6, color = "seagreen") +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "red") +
  labs(title = "Rank correlation of 4-mers",
       x = "E. coli Rank",
       y = "A. aceti Rank") +
  theme(plot.title = element_text(hjust = 0.5))
```

The scatterplot for 5-mers show the rank/rank correlation is weak which is expected as longer sequences could be species specific.

```{r}
ggplot(k5_table, aes(x = Ecoli.rank, y = Aaceti.rank)) +
  geom_point(alpha = 0.6, color = "thistle4") +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "red") +
  labs(title = "Rank correlation of 5-mers",
       x = "E. coli Rank",
       y = "A. aceti Rank") +
  theme(plot.title = element_text(hjust = 0.5))
```

###### Correlation value calculation using Spearman's ρ

Correlation values to show how strong the rank is can be done using the function `cor()` with the argument `method = "spearman"` to get **Spearman's ρ**, which is a standard measure of rank correlation [@soetewey_correlation_r]. We have already assigned ranks to the data points in our table in columns `Aaceti.rank` and `Ecoli.rank`. So then calculating the correlation between these ranks, results in a coefficient ranging from -1 to +1, where +1 is a perfect positive correlation and -1 is a perfect negative one.

```{r}
#Spearman's ρ for 3-mer
cor(k3_table$Ecoli.rank, k3_table$Aaceti.rank, method = "spearman")
#Spearman's ρ for 4-mer
cor(k4_table$Ecoli.rank, k4_table$Aaceti.rank, method = "spearman")
#Spearman's ρ for 5-mer
cor(k5_table$Ecoli.rank, k5_table$Aaceti.rank, method = "spearman")
```

###### Barplots of 10 most frequent and 10 least frequent k-mers in A. aceti

For the top 10 highest and 10 lowest frequency kmers in A. aceti, we compare the frequencies with E. coli - i.e. for the same k-mer motif. We will use a function `plot_kmer_comparison_split` to graph for each k-mer (of length 3-5).

What the below code chunk does is:

* Pull out the 10 most frequent and 10 least frequent k-mers in A. aceti.
* Show how frequent (or rare) those same k-mers are in E. coli.
* This way we directly see whether the enrichment/depletion is specific to A. aceti or shared with E. coli.

The below graphs display:

* One for the top 10 (most over-represented)
* Another for the bottom 10 (most under-represented)

```{r}
plot_kmer_comparison_split <- function(prot1, prot2, aa1, aa2, k = 3, top_n = 10,
                                       org1_name="A. aceti", org2_name="E. coli") {
  
  # Count k-mers (frequency)
  kmers1_freq <- seqinr::count(prot1, wordsize = k, alphabet = aa1, freq = TRUE)
  kmers2_freq <- seqinr::count(prot2, wordsize = k, alphabet = aa2, freq = TRUE)
  
  # Convert to data frame
  kmer_table <- data.frame(
    Kmer = names(kmers1_freq),
    Aaceti.Freq = as.numeric(kmers1_freq),
    Ecoli.Freq = as.numeric(kmers2_freq),
    stringsAsFactors = FALSE
  )
  
  # Select top and bottom k-mers in A. aceti
  over_kmers  <- head(kmer_table[order(-kmer_table$Aaceti.Freq), ], top_n)
  under_kmers <- head(kmer_table[order(kmer_table$Aaceti.Freq), ], top_n)
  
  # Overrepresented plot
  freq_matrix_over <- t(as.matrix(over_kmers[, c("Aaceti.Freq", "Ecoli.Freq")]))
  colnames(freq_matrix_over) <- over_kmers$Kmer
  
  barplot(
    freq_matrix_over,
    beside = TRUE,
    col = c("plum1", "skyblue"),
    las = 2,
    ylab = "Frequency",
    xlab = paste(k, "-mers", sep=""),
    main = paste("Top", top_n, "Over-represented", k, "-mers in", org1_name, "vs", org2_name),
    cex.names = 0.8
  )
  legend("topright", legend = c(org1_name, org2_name), fill = c("plum1", "skyblue"))
  
  # Underrepresented plot
  freq_matrix_under <- t(as.matrix(under_kmers[, c("Aaceti.Freq", "Ecoli.Freq")]))
  colnames(freq_matrix_under) <- under_kmers$Kmer
  
  barplot(
    freq_matrix_under,
    beside = TRUE,
    col = c("plum1", "skyblue"),
    las = 2,
    ylab = "Frequency",
    xlab = paste(k, "-mers", sep=""),
    main = paste("Top", top_n, "Under-represented", k, "-mers in", org1_name, "vs", org2_name),
    cex.names = 0.7
  )
  legend("topright", legend = c(org1_name, org2_name), fill = c("plum1", "skyblue"))
}

# Call plot function for kmer = 3
plot_kmer_comparison_split(aaceti_prot, ecoli_prot, aa1, aa, k=3)
# Call plot function for kmer = 4
plot_kmer_comparison_split(aaceti_prot, ecoli_prot, aa1, aa, k=4)
# Call plot function for kmer = 5
plot_kmer_comparison_split(aaceti_prot, ecoli_prot, aa1, aa, k=5)
```

##### 2.6.3. Are these k-mers also over- and under-represented in E. coli to a similar extent?

We analysed the protein sequences of A. aceti and compared them to E. coli to identify k-mers (length 3–5 amino acids) that are significantly over- or under-represented in A. aceti. Frequencies were calculated relative to the total proteome to account for differences in proteome size.

* **3-mer (ρ ~ 0.94)**
  * The ten most over-represented 3-mers in A. aceti include AAL, AAA, ALA, LAA, ALL, LLA, AGL, AAG, LAL, and LAG. 
  * Most of these k-mers are also present in E. coli but at slightly lower frequencies, indicating a species-specific enrichment. 
  * Under-represented 3-mers include CMY, CQC, KCC, MCY, WKC, WWC, YCK, CCC, CCM, and CHW, which remain rare or absent in both organisms.
  * For 3-mers, the correlation was very strong and a high correlation indicates that shorter motifs are largely conserved across the two gram-negative bacterium species.

* **4-mer (ρ ~ 0.79)**
  * Over-represented 4-mers in A. aceti include AALA, ALAA, AAAA, LAAL, ALLA, AAAL, LAAA, LLAA, AGLA, and ALAL. 
  * Many of these 4-mers show modest representation in E. coli, but the enrichment in A. aceti suggests organism-specific sequence preferences.
  * Under-represented 4-mers are mostly absent in A. aceti, while E. coli has very low but nonzero frequencies.
  * This moderate correlation suggests some divergence in sequence patterns
  
* **5-mer (ρ ~ 0.31)**
  * The top 5-mers enriched in A. aceti include GKSTL, AALAA, VSSGG, AAAAL, AAALA, AAALG, ALAAG, LAAAL, LAALR, and AARAA.
  * Many of these sequences are either absent or extremely rare in E. coli, highlighting strong organism-specific over-representation.
  * Under-represented 5-mers are largely absent in A. aceti and present only at trace levels in E. coli.
  * The weak correlation demonstrates that longer sequence contexts are highly species-specific. The rank 1 motif (GKSTL) was ranked top 1 across both organisms. However, bottom ranked sequences had very low correlation.
  
##### 2.6.4. Why do you think these sequences are present at different levels in the genomes of these organisms?

Factors influencing k-mer frequencies could be:

  * **GC content:** the genome base composition as explored in questions 1-4 showed there could be different GC content which in turn can affect amino acid usage and recurring peptides [@Agashe2014]
  * **Codon usage and amino acid frequencies:** organism specific codon bias as shown in question 5 can lead to more of certain amino acids and therefore their recurring short motifs (k-mers)
  * **Ecological niche:** adaptations to different environments may favor proteins with structural motifs that are more stable or functional under niche-specific conditions [@Hu2022]. E. coli lives in the gut, so its proteins might favor motifs aiding stress resistance and rapid growth [@DeBiase2015]. A. aceti likes acidic environments, possibly selecting for motifs stabilising proteins in low pH [@Gomes2018].
  * **Mutations and evolution over time:** over time, sequence composition may change due to selection and drift which could lead to some k-mers being amplified over others [@Bulmer1991].

In summary, these factors explain why A. aceti shows enrichment of k-mers like AAL, AAA, and AALA, while other k-mers remain rare or absent compared to E. coli. The differences in k-mer presence and frequency could reflect the unique genetic histories and functional adaptations of E. coli and A. aceti which are shaped by factors like their preferred environment and genetic content.

## Session Info

```{r}
sessionInfo()
```

## References
