---
title: "Assessment 4"
author: "Julia Lay"
date: "2025-09-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part 1: Importing files, data wrangling, mathematical operations, plots and saving code on GitHub

### Gene Expression Analysis

#### 1. Read in the file and show table with first six genes
The below reads in the file “gene_expression.tsv” making the gene identifiers the row names. The `head()` function in R then displays a table of values for the first six genes. By default, `head()` shows 6 rows but we could also specify the rows through specifying the number of rows i.e. `head(gene_data, n=6)`.

```{r}
# Read in data (tab separated)
gene_data <- read.table("gene_expression.tsv", header = TRUE, row.names = 1, sep = "\t")

# Show first six rows
head(gene_data)
```

#### 2. Add mean column

We make a new column which is the mean of the other columns. The first six genes is displayed again through the `head()` function.

```{r}
gene_data$mean <- rowMeans(gene_data)
head(gene_data)
```

#### 3. List top 10 genes with highest mean expression

The top 10 genes with the highest mean expression is shown.

```{r}
top_genes <- head(gene_data[order(-gene_data$mean), ], 10)
top_genes
```


#### 4. Determine the number of genes with a mean <10

Create variable `low_mean_count` and store the total number of mean values <10. When returning the `low_mean_count` value, we receive 35988 genes with a mean <10.

```{r}
low_mean_count <- sum(gene_data$mean < 10)
low_mean_count
```

#### 5. Make a histogram plot of the mean values and include it into your report.


```{r}
hist(gene_data$mean, breaks = 100, xlim = c(0, 550000), ylim = c(0, 60000), main = "Histogram of Mean Gene Expression", xlab = "Mean Expression")
```


### Growth Data Analysis

#### 6. Import this csv file into an R object. What are the column names?
```{r}
growth_data <- read.csv("growth_data.csv", header = TRUE)
colnames(growth_data)
```

#### 7. Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.
Sites are "northeast" or "southwest". Columns suggest start of study is 2005 and end of study is 2020

```{r}
stats <- data.frame(
  Site = unique(growth_data$Site),
  mean_2005 = tapply(growth_data$Circumf_2005_cm, growth_data$Site, mean),
  sd_2005   = tapply(growth_data$Circumf_2005_cm, growth_data$Site, sd),
  mean_2020 = tapply(growth_data$Circumf_2020_cm, growth_data$Site, mean),
  sd_2020   = tapply(growth_data$Circumf_2020_cm, growth_data$Site, sd)
)
stats
```
#### 8. Make a box plot of tree circumference at the start and end of the study at both sites.
```{r}
# Create combined vector of circumference values
circumference <- c(growth_data$Circumf_2005_cm, growth_data$Circumf_2020_cm)

# Create corresponding Site labels
site <- rep(growth_data$Site, 2)

# Create corresponding Year labels
year <- c(rep("2005", nrow(growth_data)), rep("2020", nrow(growth_data)))

# Combine Site and Year into one factor
site_year <- interaction(site, year)

# Make nicer labels
levels(site_year) <- c("Northeast 2005", "Southwest 2005",
                       "Northeast 2020", "Southwest 2020")

# Box plot
boxplot(circumference ~ site_year,
        main = "Tree Circumference at Start (2005) and End (2020)",
        xlab = "Site and Year",
        ylab = "Circumference (cm)",
        col = c("plum1", "deepskyblue", "plum1", "deepskyblue"))
```


#### 9. Calculate the mean growth over the last 10 years at each site.
It looks like 
Northeast site: 42.94 cm
Southwest site: 35.49 cm
So, on average, trees at the northeast site grew ~7.45 cm more over the last 10 years than trees at the southwest site.
```{r}
# Calculate 10-year growth per tree using 2020 and 2010
growth_data$growth10yr <- growth_data$Circumf_2020_cm - growth_data$Circumf_2010_cm

# Mean growth per site
aggregate(growth10yr ~ Site, data = growth_data, mean)
```

#### 10. Use the t.test to estimate the p-value that the 10 year growth is different at the two sites.
A Welch two-sample t-test indicated that this difference was not statistically significant (t = 1.8882, df = 87.978, p = 0.06229), with a 95% confidence interval of -0.39 to 15.29 cm. The p-value is slightly above the common significance threshold of 0.05.

```{r}
t.test(growth10yr ~ Site, data = growth_data)
```

## Part 2: Examining biological sequence diversity
As part of your assignment, you will study one of the following organisms and compare its sequence features to E. coli.

### Information
Organism allocated: Acetobacter aceti (GCA_002005445)

#### 1. Download the whole set of coding DNA sequences for E. coli and your organism of interest. How many coding sequences are present in these organisms? Present this in the form of a table. Describe any differences between the two organisms. 
Placeholder:
* Describe how you solved the problem. 
* Provide the answer or answers as directed. The answers could include numerical or categorical data, tables or charts.

Packages needed:
```{r loading packages}
suppressPackageStartupMessages({
  library("seqinr") # is a package designed to process and analyse sequence data.
  library("R.utils") # general utilities like zip and unzip
})
```

Download both E. coli (GCA_000005845) and A. aceti (GCA_002005445) from Ensembl

```{r}
#E. coli
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cds/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cds.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")
gunzip("ecoli_cds.fa.gz")

#A. aceti
URL="https://ftp.ensemblgenomes.ebi.ac.uk/pub/bacteria/release-62/fasta/bacteria_80_collection/acetobacter_aceti_gca_002005445/cds/Acetobacter_aceti_gca_002005445.ASM200544v1.cds.all.fa.gz"
download.file(URL,destfile="aaceti_cds.fa.gz")
gunzip("aaceti_cds.fa.gz")
list.files()
```
Figure out the number of coding sequences (cds) and present in a table.
```{r}
ecoli_cds <- seqinr::read.fasta("ecoli_cds.fa")
aaceti_cds <- seqinr::read.fasta("aaceti_cds.fa")

cds_table <- data.frame(
  Organism = c("E. coli", "A. aceti"),
  CDS_Count = c(length(ecoli_cds), length(aaceti_cds))
)

cds_table
```

#### 2. How much coding DNA is there in total for these two organisms? Present this in the form of a table. Describe any differences between the two organisms.

```{r}

```


#### 3. Calculate the length of all coding sequences in these two organisms. Make a boxplot of coding sequence length in these organisms. What is the mean and median coding sequence length of these two organisms? Describe any differences between the two organisms.

```{r}

```


#### 4. Calculate the frequency of DNA bases in the total coding sequences for both organisms. Perform the same calculation for the total protein sequence. Create bar plots for nucleotide and amino acid frequency. Describe any differences between the two organisms.

```{r}

```


#### 5. Create a codon usage table and quantify the codon usage bias among all coding sequences. Describe any differences between the two organisms with respect to their codon usage bias. Provide charts to support your observations.

```{r}

```


#### 6. In the organism of interest, identify 10 protein sequence k-mers of length 3-5 which are the most over- and under-represented k-mers in your organism of interest. Are these k-mers also over- and under-represented in E. coli to a similar extent? Provide plots to support your observations. Why do you think these sequences are present at different levels in the genomes of these organisms?

```{r}

```

